File: longest-line-of-consecutive-one-in-matrix/longest-line-of-consecutive-one-in-matrix.py
class Solution:
    def longestLine(self, mat: List[List[int]]) -> int:
        m = len(mat)
        n = len(mat[0])
        line = copy.deepcopy(mat)
        vertical = copy.deepcopy(mat)
        diagonal = copy.deepcopy(mat)
        antidiagonal = copy.deepcopy(mat)
        ret = mat[0][0]
        for i in range(m):
            for j in range(n):
                if line[i][j] and i:
                    line[i][j] = max(line[i][j], line[i-1][j] + 1)
                    ret = max(ret, line[i][j])
        for i in range(m):
            for j in range(n):
                if vertical[i][j] and j:
                    vertical[i][j] = max(vertical[i][j], vertical[i][j-1] + 1)
                    ret = max(ret, vertical[i][j])
        for i in range(m):
            for j in range(n):
                if diagonal[i][j] and i and j:
                    diagonal[i][j] = max(diagonal[i][j], diagonal[i-1][j-1] + 1)
                    ret = max(ret, diagonal[i][j])
        for i in range(m):
            for j in range(n):
                if antidiagonal[i][j] and i and j < n - 1:
                    antidiagonal[i][j] = max(antidiagonal[i][j], antidiagonal[i-1][j+1] + 1)
                    ret = max(ret, antidiagonal[i][j])
        return ret

File: longest-line-of-consecutive-one-in-matrix/longest-line-of-consecutive-one-in-matrix.c


int longestLine(int** mat, int matSize, int* matColSize){
    int n = *matColSize;
    int dp[matSize][n][4];
    int ret = 0;
    
    for (int i = 0; i < matSize; i++){
        for (int j = 0; j < *matColSize; j++){
            for(int k = 0; k < 4; k++){
                dp[i][j][k] = mat[i][j];
            }
            if(mat[i][j]){
                if (i)
                    dp[i][j][0] = dp[i-1][j][0] + 1;
                if (j)
                    dp[i][j][1] = dp[i][j-1][1] + 1;
                if (i && j)
                    dp[i][j][2] = dp[i-1][j-1][2] + 1;
                if (i && j < *matColSize - 1)
                    dp[i][j][3] = dp[i-1][j+1][3] + 1;
                for(int k = 0; k < 4; k++){
                    ret = fmax(ret, dp[i][j][k]);
                }
            }
        }
    }
    return ret;
}